### PR TITLE
Unify users across guilds

### PR DESCRIPTION
**Pull request by VelvetToroyashi**
This PR unifies the previous Many-to-One User:Guild relationship into a Many-to-Many relationship. While this PR is large in scale, the changes are fairly simple. Further PRs will be based off this. 

This unlocks the possibilities of tracking user activity across guilds, as well as saving on data since users are now only stored twice (once on the Users table, and once on a joiner table), instead of N times, where N is the servers that user is present on with Silk

### CODE REVIEW

File: src/Silk.Data/Entities/Guild/User/UserEntity.cs [hunk 1]
```diff
@@ -16,13 +16,15 @@ public class UserEntity
...
-    /// <summary>
-    /// The guild this user belongs to.
-    /// </summary>
-    public GuildEntity Guild { get; set; } = null!;
+    // /// <summary>
+    // /// The guild this user belongs to.
+    // /// </summary>
+    // public GuildEntity Guild { get; set; } = null!;
+
+    public ICollection<GuildEntity> Guilds { get; set; }
```
Comment (CalebABG) [inline src/Silk.Data/Entities/Guild/User/UserEntity.cs:27]: Could make type `List<GuildEntity>` but completely optional, fine as is

File: src/Silk.Data/Entities/Guild/User/UserEntity.cs [hunk 2]
```diff
@@ -33,8 +35,8 @@ public class UserEntity
     /// <summary>
     /// Non-infraction related history of the user.
     /// </summary>
-    public UserHistoryEntity History { get; set; }
+    public List<UserHistoryEntity> History { get; set; }
```
Comment (CalebABG) [inline src/Silk.Data/Entities/Guild/User/UserEntity.cs:38]: Can initialize property like below to new list 🔥

File: src/Silk.Data/MediatR/Infractions/GetUserInfractionsForGuild.cs [hunk 3]
```diff
@@ -21,12 +21,11 @@ internal sealed class Handler : IRequestHandler<Request, IEnumerable<InfractionE
...
         public async Task<IEnumerable<InfractionEntity>> Handle(Request request, CancellationToken cancellationToken)
         {
-            UserEntity? user = await _db
-                                    .Users
-                                    .Include(u => u.Infractions)
-                                    .FirstOrDefaultAsync(u => u.ID == request.TargetID && u.GuildID == request.GuildID, cancellationToken);
-
-            return user?.Infractions ?? Array.Empty<InfractionEntity>().AsEnumerable();
+            var query = _db.Infractions
+                           .Where(inf => inf.TargetID == request.TargetID)
```
Comment (CalebABG) [inline src/Silk.Data/MediatR/Infractions/GetUserInfractionsForGuild.cs:25]: Can merge the two `.Where` clauses to one: `.Where(inf => inf.TargetID == request.TargetID && inf.GuildID == request.GuildID)`

File: src/Silk.Data/Entities/Guild/User/UserEntity.cs (hunk 1 above)
Comment (VelvetToroyashi) [reply src/Silk.Data/Entities/Guild/User/UserEntity.cs:27]: Primarily used for EF Core; scared to change it. DTOs are definitely justified here.

File: src/Silk.Data/MediatR/Infractions/GetUserInfractionsForGuild.cs (hunk 3 above)
Comment (VelvetToroyashi) [reply src/Silk.Data/MediatR/Infractions/GetUserInfractionsForGuild.cs:25]: Done just for readability, but I'd be fine to change this.

**Review by CalebABG — APPROVED** [review]
LGTM 🔥

**Pull request merged**